
# *E. coli* AMR analysis workflow using Illumina data
---  

###### **_Trainers_**: [John Juma](https://github.com/ajodeh-juma), [Kennedy Mwangi](https://github.com/wanjauk), [Ouso Daniel](https://github.com/ousodaniel) & [Gilbert Kibet](https://github.com/kibet-gilbert)

---

# Set up directories
Before starting the analysis, ensure that you are logged into the HPC, create an interactive session on the assigned compute node, and change directory to the project folder which is `ACDC_AMR2025`.

```
mkdir -p \
results/illumina/ecoli/{fastqc,fastp,fastq-scan,shovill,prokka,resfinder,amrfinder,mlst,tmp/{shovill,prokka,resfinder,amrfinder}}

ln -sf /var/scratch/global/jjuma/ACDC_AMR2025/[dpsr]* .
```


# Load modules

```
module load fastqc/0.11.9
module load fastp/0.22.0
module load bbmap/38.95
module load shovill/1.1.0
module load prokka/1.14.6
module load resfinder/4.6.0
module load mlst/2.23.0
module load amrfinder/4.0.22
```

# Quality control
```
fastqc \
    -o ./results/illumina/ecoli/fastqc \
    --noextract \
    -t 2 \
    --nogroup \
    ./data/ecoli/illumina/SRR25008769_1.fastq.gz \
    ./data/ecoli/illumina/SRR25008769_2.fastq.gz
```

```
fastp \
    --in1 ./data/ecoli/illumina/SRR25008769_1.fastq.gz \
    --in2 ./data/ecoli/illumina/SRR25008769_2.fastq.gz \
    --out1 ./results/illumina/ecoli/fastp/SRR25008769_R1.trim.fastq.gz \
    --out2 ./results/illumina/ecoli/fastp/SRR25008769_R2.trim.fastq.gz \
    --detect_adapter_for_pe \
    --thread 2 \
    --json ./results/illumina/ecoli/fastp/SRR25008769.fastp.json \
    --html ./results/illumina/ecoli/fastp/SRR25008769.fastp.html \
    --cut_mean_quality 20 \
    --cut_front \
    --cut_tail \
    --cut_window_size 4 \
    --qualified_quality_phred 25 \
    --unqualified_percent_limit 40 \
    --length_required 20 \
    2>&1 | tee ./results/illumina/ecoli/fastp/SRR25008769.fastp.log
```

# De novo assembly pipeline for Illumina paired reads
Assemble bacterial isolate genomes from Illumina paired-end reads

Shovill is a pipeline which uses SPAdes at its core, but alters the steps before
and after the primary assembly step to get similar results in less time. Shovill
also supports other assemblers like SKESA, Velvet and Megahit, so you can take
advantage of the pre- and post-processing the Shovill provides with those too.


>**Note**
Shovill is for isolate data only, primarily small haploid organisms. 
It will NOT work on metagenomes or larger genomes. 
Please use Megahit directly instead.


# Main steps

1. Estimate genome size and read length from reads (unless --gsize provided) - mash
2. Reduce FASTQ files to a sensible depth (default --depth 100) - seqtk
3. Trim adapters from reads (with --trim only) - trimmomatic
4. Conservatively correct sequencing errors in reads - lighter
5. Pre-overlap ("stitch") paired-end reads - flash
6. Assemble with SPAdes/SKESA/Megahit with modified kmer range and PE + long SE reads
7. Correct minor assembly errors by mapping reads back to contigs (bwa-mem + pilon)
8. Remove contigs that are too short, too low coverage, or pure homopolymers
9. Produce final FASTA with nicer names and parseable annotations



```
export TMPDIR="./results/illumina/ecoli/tmp/shovill"
```

```
shovill \
  --R1 ./results/illumina/ecoli/fastp/SRR25008769_R1.trim.fastq.gz \
  --R2 ./results/illumina/ecoli/fastp/SRR25008769_R2.trim.fastq.gz \
  --gsize 5249449 \
  --outdir ./results/illumina/ecoli/shovill/SRR25008769 \
  --assembler skesa \
  --minlen 500 \
  --mincov 2 \
  --force \
  --keepfiles \
  --depth 0 \
  --noreadcorr \
  --namefmt "SRR25008769_%05d" \
  --cpus 4 \
  --ram 16 \
  --tmpdir $TMPDIR
```

```
mv ./results/illumina/ecoli/shovill/SRR25008769/contigs.fa ./results/illumina/ecoli/shovill/SRR25008769/SRR25008769.fa
```

# Assembly evaluation

```
stats.sh in=./results/illumina/ecoli/shovill/SRR25008769/SRR25008769.fa
```

>**Note**
Unfortunately, the N50 and L50 values generated by stats.sh are switched. N50
should be a length and L50 should be a count. The results table below shows the
corrected values based on stats.sh outputs.


# Annotation
```
export TMPDIR=./results/illumina/ecoli/tmp/prokka/
```

```
prokka \
    --evalue 1e-09 \
    --centre ILRI \
    --coverage 80 \
    --cpus 2 \
    --prefix SRR25008769 \
    --locustag SRR25008769 \
    --proteins ./databases/prokka/proteins.faa \
    --force \
    --outdir ./results/illumina/ecoli/prokka \
    ./results/illumina/ecoli/shovill/SRR25008769/SRR25008769.fa
```

# AMR genes detection using ResFinder

```
python -m resfinder \
    -ifa ./results/illumina/ecoli/shovill/SRR25008769/SRR25008769.fa \
    -o ./results/illumina/ecoli/resfinder/ecoli \
    -s ecoli \
    --min_cov 0.6 \
    --threshold 0.9 \
    --min_cov_point 0.6 \
    --threshold_point 0.9 \
    --ignore_stop_codons \
    --ignore_indels \
    --acquired \
    --point
```

# Full AMRFinderPlus search combining results

Identify acquired antimicrobial resistance genes in bacterial protein and/or assembled nucleotide sequences as well as known resistance-associated point mutations for several taxa. With AMRFinderPlus we added select members of additional classes of genes such as virulence factors, biocide, heat, acid, and metal resistance genes.

>**Note**
AMRFinderPlus reports gene and point mutation presence/absence; it does not
infer phenotypic resistance. Many of the resistance genes detected by
AMRFinderPlus may not be relevant for clinical management or antimicrobial
surveillance. **The AMR genes must be expressed to confer resistance**

AMRFinder can be run in multiple modes with protein sequence as input and/or with DNA sequence as input. To get maximum information it should be run with both protein and nucleotide. When run with protein sequence it uses both BLASTP and HMMER to search protein sequences for AMR genes along with a hierarchical tree of gene families to classify and name novel sequences. With nucleotide sequences it uses BLASTX translated searches and the hierarchical tree of gene families. Adding the `--organism` option enables screening for point mutations in select organisms and suppresses the reporting of some that are extremely common in those organisms.


```
export TMPDIR=./results/illumina/ecoli/tmp/amrfinder/
```

```
AMRFINDER_DB=$(find /export/apps/amrfinder/4.0.22/data/2025-03-25.1 -name "AMR.LIB" | sed 's=AMR.LIB==')
```

```
amrfinder \
    --nucleotide ./results/illumina/ecoli/prokka/SRR25008769.fna \
    --protein ./results/illumina/ecoli/prokka/SRR25008769.faa \
    --gff ./results/illumina/ecoli/prokka/SRR25008769.gff \
    --annotation_format prokka \
    --organism Escherichia \
    --plus \
    --ident_min -1 \
    --coverage_min 0.5 \
    --translation_table 11 \
    --database $AMRFINDER_DB \
    --threads 2 \
    --name SRR25008769 > ./results/illumina/ecoli/amrfinder/SRR25008769.tsv
```

# Multilocus sequence typing

```
MLST_DB=$(find ./databases/mlst/database/ -name "mlst.fa" | sed 's=blast/mlst.fa==')
```

```
mlst \
    --threads 2 \
    --blastdb $MLST_DB/blast/mlst.fa \
    --datadir $MLST_DB/pubmlst \
    --scheme ecoli_achtman_4 \
    --minid 95 \
    --mincov 10 \
    --minscore 50 \
    ./results/illumina/ecoli/prokka/SRR25008769.fna \
    > ./results/illumina/ecoli/mlst/SRR25008769.tsv
```

